powertools-idempotency/src/test/java/software/amazon/lambda/powertools/idempotency/internal/IdempotencyAspectTest.java

```diff
@@ -0,0 +1,270 @@
+        assertThatThrownBy(() -> function.handleRequest(p, context)).isInstanceOf(IdempotencyAlreadyInProgressException.class);
+    }
+
+    @Test
+    public void functionThrowException_shouldDeleteRecord_andThrowFunctionException() {
+        // GIVEN
+        Idempotency.config()
+                .withPersistenceStore(store)
+                .withConfig(IdempotencyConfig.builder()
+                        .withEventKeyJMESPath("id")
+                        .build()
+                ).configure();
+
+        // WHEN / THEN
+        IdempotencyWithErrorFunction function = new IdempotencyWithErrorFunction();
+
+        Product p = new Product(42, "fake product", 12);
+        assertThatThrownBy(() -> function.handleRequest(p, context))
+                .isInstanceOf(IndexOutOfBoundsException.class);
+
+        verify(store).deleteRecord(any(), any(IndexOutOfBoundsException.class));
+    }
+
+    @Test
+    @SetEnvironmentVariable(key = Constants.IDEMPOTENCY_DISABLED_ENV, value = "true")
+    public void testIdempotencyDisabled_shouldJustRunTheFunction() {
+        // GIVEN
+        Idempotency.config()
+                .withPersistenceStore(store)
+                .withConfig(IdempotencyConfig.builder()
+                        .withEventKeyJMESPath("id")
+                        .build()
+                ).configure();
+
+        // WHEN
+        IdempotencyEnabledFunction function = new IdempotencyEnabledFunction();
+        Product p = new Product(42, "fake product", 12);
+        Basket basket = function.handleRequest(p, context);
+
+        // THEN
+        verifyNoInteractions(store);
+        assertThat(basket.getProducts()).hasSize(1);
+        assertThat(function.handlerCalled()).isTrue();
+    }
+
+    @Test
+    public void idempotencyOnSubMethodAnnotated_firstCall_shouldPutInStore() {
+        Idempotency.config()
+                .withPersistenceStore(store)
+                .withConfig(IdempotencyConfig.builder().build()
+                ).configure();
+
+        // WHEN
+        IdempotencyInternalFunction function = new IdempotencyInternalFunction();
+        Product p = new Product(42, "fake product", 12);
+        Basket basket = function.handleRequest(p, context);
+
+        // THEN
+        assertThat(basket.getProducts()).hasSize(2);
+        assertThat(function.subMethodCalled()).isTrue();
+
+        ArgumentCaptor<JsonNode> nodeCaptor = ArgumentCaptor.forClass(JsonNode.class);
+        verify(store).saveInProgress(nodeCaptor.capture(), any());
+        assertThat(nodeCaptor.getValue().asText()).isEqualTo("fake");
+
+        ArgumentCaptor<Basket> resultCaptor = ArgumentCaptor.forClass(Basket.class);
+        verify(store).saveSuccess(any(), resultCaptor.capture(), any());
+        assertThat(resultCaptor.getValue().getProducts()).contains(basket.getProducts().get(0));
+        assertThat(resultCaptor.getValue().getProducts()).contains(new Product(0, "fake", 0));
```

Comment:
Can be combined into one assert ?

```suggestion
         assertThat(resultCaptor.getValue().getProducts())
                .contains(basket.getProducts().get(0), new Product(0, "fake", 0));
```